streams/src/main/java/org/apache/kafka/streams/processor/internals/ChangelogReader.java

```diff
@@ -21,29 +21,35 @@
 
 import java.util.Collection;
 import java.util.Map;
+import java.util.Set;
 
 /**
  * Performs bulk read operations from a set of partitions. Used to
- * restore  {@link org.apache.kafka.streams.processor.StateStore}s from their
- * changelogs
+ * restore a state store from {@link TopicPartition} changelog.
  */
 public interface ChangelogReader {
     /**
      * Register a state store and it's partition for later restoration.
-     * @param restorer the state restorer to register
+     *
+     * @param partition the changelog topic partition to restore
+     * @param stateManager the state manager used for restoring (one per task)
      */
-    void register(final StateRestorer restorer);
+    void register(final TopicPartition partition, final ProcessorStateManager stateManager);
 
     /**
      * Restore all registered state stores by reading from their changelogs.
-     * @return all topic partitions that have been restored
      */
-    Collection<TopicPartition> restore(final RestoringTasks active);
+    void restore();
 
     /**
-     * @return the restored offsets for all persistent stores.
+     * Update offset limit of a given changelog partition
      */
-    Map<TopicPartition, Long> restoredOffsets();
+    void updateLimitOffsets();
```

Comment:
This is to be called by the task manager when it wants to update the limit offsets: here some optimization can be considered e.g. we know that limit offset may be updated when the thread is about to commit, so maybe other thread hosting the active tasks of its standbys may also be committing.

Reply:
The javadocs imply this should accept a changelog partition, is it missing an argument or are the docs incorrect?

Reply:
The docs are incorrect now: the partitions are managed inside the changelog reader, and when it is called the changelog reader knows which changelogs it maintains has a limit offset that can be updated -- i.e. the caller only need to consider the calling frequency so to speak.

Thanks for the catch.